internal/integration/entity_synthesis.go

```diff
@@ -0,0 +1,172 @@
+// Package integration ...
+// Copyright 2021 New Relic Corporation. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+package integration
+
+import (
+	"sort"
+	"strings"
+
+	sdk_metadata "github.com/newrelic/infra-integrations-sdk/v4/data/metadata"
+	"github.com/newrelic/nri-prometheus/internal/pkg/labels"
+)
+
+// Synthesizer group of rules to synthesis entities
+type Synthesizer struct {
+	rulesByConditions map[Condition]EntityRule
+}
+
+// NewSynthesizer initialize and return a Synthesizer from a set of EntityRules
+func NewSynthesizer(entitySynthesisDefinitions []SynthesisDefinition) Synthesizer {
+	s := Synthesizer{
+		rulesByConditions: make(map[Condition]EntityRule),
+	}
+	for _, ed := range entitySynthesisDefinitions {
+		ed.tagRules = ed.Tags.getTagRules()
+		s.addConditions(ed.EntityRule)
+
+		for _, er := range ed.Rules {
+			er.EntityType = ed.EntityType
+			er.tagRules = er.Tags.getTagRules()
+			er.tagRules = append(er.tagRules, ed.EntityRule.tagRules...)
+			s.addConditions(er)
+		}
+	}
+	return s
+}
+func (s *Synthesizer) addConditions(rule EntityRule) {
+	if rule.Identifier == "" || rule.Name == "" || rule.EntityType == "" {
+		return
+	}
+	for _, c := range rule.Conditions {
+		s.rulesByConditions[c] = rule
+	}
+}
+
+// SynthesisDefinition contains rules to synthesis entities from metrics
+type SynthesisDefinition struct {
+	EntityRule `mapstructure:",squash"`
+	Rules      []EntityRule `mapstructure:"rules"`
+}
+
+// EntityRule contains rules to synthesis an entity
+type EntityRule struct {
+	EntityType string      `mapstructure:"type"`
+	Identifier string      `mapstructure:"identifier"`
+	Name       string      `mapstructure:"name"`
+	Conditions []Condition `mapstructure:"conditions"`
+	Tags       Tags        `mapstructure:"tags"`
+	tagRules   []tagRule
+}
+
+// Condition contains parameters used to match entities from metrics
+type Condition struct {
+	Attribute string `mapstructure:"attribute"`
+	Prefix    string `mapstructure:"prefix"`
+	Value     string `mapstructure:"value"`
+}
+
+func (c Condition) match(attribute string) bool {
```

Comment:
I guess we could remove the second ìf`. because if c.Prefix==""  the has.Prefix will return true always.

```go
func (c Condition) match(attribute string) bool {
	if c.Value != "" {
		return c.Value == attribute
	}
	if c.Prefix != "" {
		return strings.HasPrefix(attribute, c.Prefix)
	}
	// if Value and Prefix are empty there is a match since the attribute exists
	return true
}

func (c Condition) match2(attribute string) bool {
	if c.Value != "" {
		return c.Value == attribute
	}
	return strings.HasPrefix(attribute, c.Prefix)
}
```

Reply:
I think it will be possible , and i will leave the same comment to clarify the intention behind the logic